dilated-file-transfer.md

```diff
@@ -0,0 +1,476 @@
+(XXX see "file naming" in discussion)
+
+
+## Version Negotiation
+
+There is an existing file-transfer protocol which does not use Dilation (called "classic" in this document).
+Clients supporting newer versions of file-transfer (i.e. the one in this document) SHOULD offer backwards compatibility where possible.
+
+In the core mailbox protocol, applications can indicate version information, via `app_versions`.
+The existing file-transfer protocol doesn't use this feature so the version information is empty (indicating "classic").
+This new Dilated File Transfer MUST include version information:
+
+```json
+{
+    "transfer": {
+        "mode": "{send|receive|connect}",
+        "features": ["core0"],
+        "permission": "{ask|yes}"
+    }
+}
+```
+
+**Rejected idea**: having a `version` number that increases.
+Per RFC 9170, it can be the case that a protocol can "ossify" or lost its flexibility, including when using "highest common version" sorts of negotiation.
+Multiple extension points (e.g. both "version" and "features") can cause confusion; including both was rejected after considering the question, "when would `version` be incremented instead of using a `feature`?"
+
+The `mode` key indicates the desired mode of that peer.
+It has one of three values:
+* `"send"`: the peer will only send files (similar to classic transfer protocol)
+* `"receive"`: the peer only receive files (the flip side of the above)
+* `"connect"`: the peer will send and receive zero or more files before closing the session
+
+Note that `send` and `receive` above will still use Dilation as all clients supporting this protocol must.
+If a peer sends no version information at all, it will be using the classic protocol (and is thus using Transit and not Dilation for the peer-to-peer connection).
+
+The `"features"` key is a list of message-formats / features understood by the peer.
+This allows for existing messages to be extended, or for new message types to be added.
+Peers MUST _accept_ messages for any features they support.
+Peers MUST only send messages for features in the other side's list.
+Only one feature exists currently: `"core0"` (and it MUST be supported).
+Peers MUST tolerate the existence of unknown values in the `"features"` list.
+
+   (XXX might make the Python implementation randomly add an unknown one, 10% of the time?)
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes or capabilities or evolving existing ones.
+
+The `"permission"` key specifies how to proceed with offers.
+Using mode `ask` tells the sender to pause after transmitting the first metadata message and await an answer from the peer before streaming data.
+Using mode `yes` means the peer will accept all incoming transfers so the sender should not pause after the metadata (and instead immediately start sending data messages).
+This cuts down latency for "one-way" transfers (see "Discussion")
```

Comment:
So, this isn't only during the initial connection establishment -- it's for every subchannel (remember there is one subchannel per Offer).

That said, it may be unneccesary complication (as indeed it's only one round-trip).

This came from (now somewhat old) original discussion with Brian Warner about his ideal UX etc -- the use-case here being that a single human controls two computers, and transfers between them: they might want an "accept" mode on the receiving computer so they can just keep dropping files into the sending said.

All that said, this "UX treatment" can still be accomplished with an "ask, then send data" mode (only) since the software can just automatically answer "yes".

I have been continuing the prototype branch in Python and having this "mode" does make the state-machine quite a bit larger -- so perhaps it _should_ be relegated to some future extension in the interests of simplicity (since the only gain is, as you point out, some dubious efficiency-gains -- one more round-trip per Offer).